### PR TITLE
[Experiment] Additional field extensible sanitisation and validation handling

### LINKED ISSUE #44019: Validate the additional checkout fields against custom validation hooks and field schema when editing in My Account area
When a shopper goes to the My Account area and edits the additional checkout fields registered in the `address` area, we do not run any validation on the values. This is not a major concern as the shopper would not be able to finish the checkout process with the invalid values, however for a nicer UX we should try to catch this in the my account area too.

### PR DESCRIPTION
**Pull request by mikejolley**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Introduces new hooks for extensions to sanitise and validate registered checkout fields.

- Validation now uses an action hook with the error object passed by reference. This negates the need to compare the error object before and after hooks to ensure the object did not change
- Add sanitisation hook for custom sanitisation rules on additional fields
- Validation hook for custom validation rules on additional fields - key and value only
- Location validation hook for custom validation rules on additional fields where there are dependencies on the values of other fields in that location
- Validation applied to my account area

Closes #44019 

### How to test the changes in this Pull Request:

Developers should use this snippet this add some custom fields. Gov ID and email have validation methods.

<details>

```
add_action(
	'woocommerce_blocks_loaded',
	function() {
		woocommerce_blocks_register_checkout_field(
			array(
				'id'                => 'plugin-namespace/alt-email',
				'label'             => 'Alternative Email Field',
				'location'          => 'contact',
				'type'              => 'text',
				'required'          => true,
				'sanitize_callback' => function( $field_value ) {
					return sanitize_email( $field_value );
				},
				'validate_callback' => function( $field_value ) {
					if ( ! is_email( $field_value ) ) {
						return new \WP_Error( 'invalid_alt_email', 'Please ensure your alternative email matches the correct format.' );
					}
				},
			),
		);
		woocommerce_blocks_register_checkout_field(
			array(
				'id'                => 'plugin-namespace/gov-id',
				'label'             => 'Government ID',
				'location'          => 'address',
				'type'              => 'text',
				'required'          => true,
				'sanitize_callback' => function( $field_value ) {
					return str_replace( ' ', '', $field_value );
				},
			),
		);
		woocommerce_blocks_register_checkout_field(
			array(
				'id'       => 'plugin-namespace/confirm-gov-id',
				'label'    => 'Confirm Government ID',
				'location' => 'address',
				'type'     => 'text',
				'required' => true,
			),
		);
		woocommerce_blocks_register_checkout_field(
			array(
				'id'       => 'plugin-namespace/leave-on-porch',
				'label'    => __( 'Please leave my package on the porch if I\'m not home', 'woocommerce' ),
				'location' => 'additional',
				'type'     => 'checkbox',
				'required' => true,
			),
		);
		woocommerce_blocks_register_checkout_field(
			array(
				'id'       => 'plugin-namespace/location-on-porch',
				'label'    => __( 'Describe where we should hide the parcel', 'woocommerce' ),
				'location' => 'additional',
				'type'     => 'text',
			)
		);
		woocommerce_blocks_register_checkout_field(
			array(
				'id'       => 'plugin-namespace/leave-with-neighbor',
				'label'    => __( 'Which neighbor should we leave it with if unable to hide?', 'woocommerce' ),
				'location' => 'additional',
				'type'     => 'select',
				'options'  => array(
					array(
						'label' => 'Neighbor to the left',
						'value' => 'left',
					),
					array(
						'label' => 'Neighbor to the right',
						'value' => 'right',
					),
					array(
						'label' => 'Neighbor across the road',
						'value' => 'across',
					),
					array(
						'label' => 'Do not leave with a neighbor',
						'value' => 'none',
					),
				),
			)
		);

		add_action(
			'woocommerce_blocks_sanitize_additional_field',
			function( $value, $key ) {
				if ( 'plugin-namespace/confirm-gov-id' === $key ) {
					return str_replace( ' ', '', $value );
				}
				return $value;
			},
			10,
			2
		);
		add_action(
			'woocommerce_blocks_validate_additional_field',
			function ( \WP_Error $errors, $field_key, $field_value ) {
				if ( 'plugin-namespace/gov-id' === $field_key ) {
					$match = preg_match( '/[A-Z0-9]{5}/', $field_value );
					if ( 0 === $match || false === $match ) {
						$errors->add( 'invalid_gov_id', 'Please ensure your government ID matches the correct format.' );
					}
				}
			},
			10,
			3
		);
		add_action(
			'woocommerce_blocks_validate_location_address_fields',
			function ( \WP_Error $errors, $fields, $group ) {
				if ( $fields['plugin-namespace/gov-id'] !== $fields['plugin-namespace/confirm-gov-id'] ) {
					$errors->add( 'gov_id_mismatch', 'Please ensure your government ID matches the confirmation.' );
				}
			},
			10,
			3
		);
	}
);
```

</details>

1. Add items to cart and go to checkout.
2. Enter something in the custom "additional email" field that is not an email, and enter some letters in the Gov ID fields. Ensure both Gov ID address fields are different.
3. After typing in Gov ID the checkout will refresh. You'll see an error notice with 2 bullets (formatting is wrong, and fields do not match). Correct this and ensure the errors go away.
4. Place order and it will throw another error this time about email. Correct this and place the order again. It should go through.
5. Check the confirmation page values match what you submitted.
6. Go to the My Account page.
7. Under addresses, edit an address. Try changing the Gov ID fields and submit with errors. You'll see an error notice if it fails validation.
8. If you refresh the page before fixing the validation errors, correct values are restored.
9. If you fix the validation errors, the new values should save to the account.
10. Under "account details" change "Alternative Email" to something else. Save. See error message.
11. Correct the email field and save. Change should persist.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFields.php [hunk 1]
```diff
@@ -537,73 +538,84 @@ public function get_field_location( $field_key ) {
...
+					'The action %s encountered an error. The field %s will not have any custom validation applied to it. %s',
+					'woocommerce_blocks_validate_additional_field',
+					esc_html( $field_key ),
 					esc_html( $e->getMessage() )
 				),
 				E_USER_WARNING
 			);
 
-			return new \WP_Error();
+			return new WP_Error();
```
Comment (senadir) [inline plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFields.php]: Why not return errors? suspect it's faulty?

File: plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFields.php [hunk 2]
```diff
@@ -688,17 +682,62 @@ function( $key ) use ( $order_fields_keys ) {
...
+	public function validate_fields_for_location( $fields, $location, $group = '' ) {
+		$errors = new WP_Error();
+
+		try {
+			/**
+			 * Pass an error object to allow validation of an additional field.
+			 *
+			 * @param WP_Error $errors      A WP_Error object that extensions may add errors to.
+			 * @param mixed    $field_value The value of the field being validated.
+			 * @param string   $field_key   Key of the field being sanitized.
```
Comment (opr) [inline plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFields.php]: Params should be updated here

File: plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php [hunk 3]
```diff
@@ -426,10 +427,9 @@ private function update_customer_from_request( \WP_REST_Request $request ) {
 		$shipping_address_values = $request['shipping_address'] ?? $request['billing_address'];
 
 		foreach ( $shipping_address_values as $key => $value ) {
-			if ( is_callable( [ $customer, "set_shipping_$key" ] ) ) {
-				$customer->{"set_shipping_$key"}( $value );
-			} elseif ( 'phone' === $key ) {
```
Comment (mikejolley) [inline plugins/woocommerce/src/StoreApi/Routes/V1/Checkout.php:431]: Removed on purpose; core has a set_shipping_phone method now.

**Review by opr — COMMENTED** [review]
Tested this out and seems to be working fine - I did look into the code and couldn't spot anything that should be actioned on this PR (left you a question on Slack for a follow up)

I will revisit on Monday and try some more edge cases too.

One other thing to mention is the order of arguments on the new hooks you added. Their args go `$value` first then `$key` second. I'd suggest making the key first and value second, but that's just my personal opinion, I am more used to saying "key/value" so that's what seems more natural to me.

If you decide not to change it then maybe consider updating the args of `validate_field` to match the order in the hook, just for consistency and to make it a little easier for future us to reason with.

Good work!

**Comment by mikejolley** [comment]
> One other thing to mention is the order of arguments on the new hooks you added. Their args go $value first then $key second. I'd suggest making the key first and value second, but that's just my personal opinion, I am more used to saying "key/value" so that's what seems more natural to me.

Usually it depends if its the value thats the focus of the function/hook to what comes first. If its a hook, and we're filtering value, value must come first. Unless you spotted more places?

**Comment by opr** [comment]
> > One other thing to mention is the order of arguments on the new hooks you added. Their args go $value first then $key second. I'd suggest making the key first and value second, but that's just my personal opinion, I am more used to saying "key/value" so that's what seems more natural to me.
> 
> Usually it depends if its the value thats the focus of the function/hook to what comes first. If its a hook, and we're filtering value, value must come first. Unless you spotted more places? 
> 
> 

Yeah I'm fine with either I just think `validate_field` should match then

**Review by nielslange — COMMENTED** [review]
The 2nd and 3rd testing instructions say

> 2. Enter something in the custom "additional email" field that is not an email, and enter some letters in the Gov ID fields. Ensure both Gov ID address fields are different.
> 3. After typing in Gov ID the checkout will refresh. You'll see an error notice with 2 bullets (formatting is wrong, and fields do not match). Correct this and ensure the errors go away.

On my end, I'm only seeing the following error notice with a single error message and no bullet points:

Screenshot 2024-02-12 at 13 44 57

**Review by nielslange — APPROVED** [review]
Thanks for working on this, @mikejolley. I've successfully tested the PR. While checking the code and the code snippet, I noticed that the validation hooks might have inconsistent names. 

In the code snippet, I see the following two hooks:
- `woocommerce_blocks_validate_additional_field`
- `woocommerce_blocks_validate_address_fields`

I noticed that `woocommerce_blocks_validate_additional_field` validates both the custom fields in the "Contact information" section and the one in the "Additional order information" section. 

Furthermore, I noticed that `woocommerce_blocks_validate_additional_field` is using the singular (`[...]_field`), while `woocommerce_blocks_validate_address_fields` is using the plural (`[...]_fields`).

I wonder if we should stick to either singular or plural for consistency reasons. I also wonder if we should have individual validation hooks (e.g. `woocommerce_blocks_validate_contact_fields`, `woocommerce_blocks_validate_address_fields` and `woocommerce_blocks_validate_additional_fields`) or one general validation hook (e.g. `woocommerce_blocks_validate_custom_fields`).

As this PR works as expected, I'm approving it. As for the validation hooks, I'll leave this up to you and the extensibility squad to decide if it's worth renaming them, introducing one general validation hook or creating individual hooks for each section. Or simply keep the validation as it is. I might have missed earlier discussions about that.

**Comment by opr** [comment]
> I noticed that woocommerce_blocks_validate_additional_field validates both the custom fields in the "Contact information" section and the one in the "Additional order information" section.
>
> Furthermore, I noticed that woocommerce_blocks_validate_additional_field is using the singular ([...]_field), while woocommerce_blocks_validate_address_fields is using the plural ([...]_fields).

To clarify `woocommerce_blocks_validate_additional_field` will be called for each field no matter where it is registered (contact, address, additional) and it is only intended to cover a single field (it is called multiple times, once for each field).

`woocommerce_blocks_validate_address_fields` will run once per location, and information about each field in that location will be passed.

The difference is how many fields each hook is responsible for validating, so I think the plural/singular use here is fine.

File: plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFields.php [hunk 4]
```diff
@@ -688,17 +680,60 @@ function( $key ) use ( $order_fields_keys ) {
...
+			/**
+			 * Pass an error object to allow validation of an additional field.
+			 *
+			 * @param WP_Error $errors  A WP_Error object that extensions may add errors to.
+			 * @param mixed    $fields  List of fields (key value pairs) in this location.
+			 * @param string   $group   The group of this location (shipping|billing|'').
+			 *
+			 * @since 8.7.0
+			 */
+			do_action( 'woocommerce_blocks_validate_' . $location . '_fields', $errors, $fields, $group );
```
Comment (opr) [inline plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFields.php]: When the `additional` area is validated, the hook is:
`woocommerce_blocks_validate_additional_fields` which may be easily confused with `woocommerce_blocks_validate_additional_field`. Do you think this is an issue? Should we consider renaming one or the other? Maybe adding `location` to the location hook would work? `woocommerce_blocks_validate_location_additional_fields`

Comment (mikejolley) [reply plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFields.php]: Added in https://github.com/woocommerce/woocommerce/pull/44463/commits/b42414966ced846d1166f8874a66a5f37a73be50

**Comment by mikejolley** [comment]
Good now @opr ? Ill get tests passing again.

**Review by opr — APPROVED** [review]
Working nicely now, thanks!

We should ensure sanitization is also tested, here's the example code I used:

```php
add_action(
	'woocommerce_blocks_sanitize_additional_field',
	function( $value, $key ) {
		if ( 'plugin-namespace/gov-id' === $key || 'plugin-namespace/confirm-gov-id' === $key ) {
			return str_replace( ' ', '', $value );
		}
		return $value;
	},
	10,
	2
);
```

feel free to add that to the instructions if you like.

**Review by opr — APPROVED** [review]
Thanks Mike, this is working great. My errors are being caught, checkout can continue 🙌🏼

**Pull request merged**